fix(transformer/async-to-generator): move parameters to the inner generator function when they could throw errors

Dunqing opened the pull request:
The new implementation port from [esbuild](https://github.com/evanw/esbuild/blob/df815ac27b84f8b34374c9182a93c94718f8a630/internal/js_parser/js_parser_lower.go#L355-L467), before from `Babel`. 

Babel's transform implementation for the async method is incorrect because the async method should return a rejecting promise when it throws an error. Everything is good if the errors are thrown in the async method body, but the following case will throw an error in the parameters which causes the whole program crushed not a rejecting promise. So we should move the parameters to the inner generator function when the parameters could throw an error.

Input: 
```js
class Cls {
  // ReferenceError: Cannot access 'b' before initialization
  async method(a = b, b = 0) {}
}
```

Before output
```js
class Cls {
  method(a = b, b = 0) {
    return babelHelpers.asyncToGenerator(function* () {})();
  }
}
```

After output:
```js
class Cls {
  method() {
     // ReferenceError: Cannot access 'b' before initialization
    return babelHelpers.asyncToGenerator(function* (a = b, b = 0) {}).apply(this, arguments);
  }
}
```


No override tests because Babel doesn't cover this case.

Dunqing commented:
* **#8500** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8500?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8500?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8500).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Dunqing commented:
I noticed a potential improvement, we don't need to change `this` to `_this` in the async method and async arrow function because both `asyncToGenerator` and `wrapAsyncGenerator` are binding the `this` for the inner generator function. However this will cause a lot of tests to fail, thus we need to override them, this is not a good moment to change this.

overlookmotel reviewed (approved):
Nice!

overlookmotel commented:
### Merge activity

* **Jan 15, 7:01 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
fix(transformer/async-to-generator): move parameters to the inner generator function when they could throw errors (#8500)

The new implementation port from [esbuild](https://github.com/evanw/esbuild/blob/df815ac27b84f8b34374c9182a93c94718f8a630/internal/js_parser/js_parser_lower.go#L355-L467), before from `Babel`.

Babel's transform implementation for the async method is incorrect because the async method should return a rejecting promise when it throws an error. Everything is good if the errors are thrown in the async method body, but the following case will throw an error in the parameters which causes the whole program crushed not a rejecting promise. So we should move the parameters to the inner generator function when the parameters could throw an error.

Input:
```js
class Cls {
  // ReferenceError: Cannot access 'b' before initialization
  async method(a = b, b = 0) {}
}
```

Before output
```js
class Cls {
  method(a = b, b = 0) {
    return babelHelpers.asyncToGenerator(function* () {})();
  }
}
```

After output:
```js
class Cls {
  method() {
     // ReferenceError: Cannot access 'b' before initialization
    return babelHelpers.asyncToGenerator(function* (a = b, b = 0) {}).apply(this, arguments);
  }
}
```

No override tests because Babel doesn't cover this case.

overlookmotel commented:
One things for follow-up: There are more things beyond `AssignmentPattern` which can throw. e.g.:

```js
const throwIt = () => { throw new Error('gotcha!'); };

const constant = 1;

Object.defineProperty(global, 'foo', { set(v) {} }); // No getter

class C {
    async method1( [a = b], b = 0 ) {}
    async method2( {x: a = b}, b = 0 ) {}
    async method3( { [throwIt()]: a } ) {}
    async method4( { [constant = 2]: a } ) {}
    async method5( { [foo]: a } ) {} // Just trying to read `global.foo` throws an error!
    async method6( a = foo ) {} // Ditto
}
```

It might be best to conservatively assume that any computed keys in object destructuring can throw - they'll be rare anyway.

Any accesses of unbound variables can also throw, and no way to know statically if they will or not.

overlookmotel commented:
Force pushed to resolve merge conflicts in snapshots.
